MoreLinq.Test/MemoizeTest.cs

```diff
@@ -0,0 +1,129 @@
+﻿#region License and Terms
+// MoreLINQ - Extensions to LINQ to Objects
+// Copyright (c) 2016 Leandro F. Vieira (leandromoh). All rights reserved.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#endregion
+
+using System;
+using System.Collections.Generic;
+using NUnit.Framework;
+using System.Linq;
+
+namespace MoreLinq.Test
+{
+    [TestFixture]
+    public class MemoizeTest
```

Comment:
I'm missing a test that ensures that a source sequence is enumerated exactly once:

```csharp

[Test]
public void MemoizedIsEnumeratedOnlyOnce() {
     var src = new SingleUseEnumerable<int>(new [] {1,2,3,4}).Memoize();
     src.ToList();
     src.ToList();
}

private class SingleUseEnumerable<T> : IEnumerable<T>{
     private readonly IEnumerable<T> src;
     private bool enumerated = false;
     public SingleUseEnumerable(IEnumerable<T> src) {
        this.src = src;
     }

     public IEnumerator<T> GetEnumerator() {
         if (enumerated) throw new InvalidOperationException("enumerated twice");

         foreach(var i in src) yield return i;
         enumerated = true;
    }
}
```